src/client/dfs.cc

```diff
@@ -1151,46 +693,337 @@ bool DFS::exists(std::string name) {
+
+  //! Send file to be submitted;
+  auto socket = connect(file_hash_key);
+  send_message(socket.get(), &file_info);
+
+  //! Get information of where to send the file
+  auto description = read_reply<FileDescription>(socket.get());
+  socket->close();
+
+  IOoperation io_ops;
+  io_ops.operation = eclipse::messages::IOoperation::OpType::BLOCK_INSERT;
+
+  int which_server = (description->hash_key % NUM_NODES) % NUM_NODES;
+
+  BlockMetadata metadata;
+  metadata.name = file_name + "_" + to_string(0);
+  metadata.file_name = file_name;
+  metadata.hash_key = boundaries.random_within_boundaries(which_server);
+  metadata.seq = 0;
+  metadata.size = 0;
+  metadata.type = static_cast<unsigned int>(type);
+  metadata.replica = replica;
+  metadata.node = nodes[which_server];
+  metadata.l_node = nodes[(which_server - 1 + NUM_NODES) % NUM_NODES];
+  metadata.r_node = nodes[(which_server + 1 + NUM_NODES) % NUM_NODES];
+  metadata.is_committed = 1;
+
+  Block block;
+  block.first = metadata.name;
+  block.second = "";
+
+  io_ops.block = move(block);
+
+  socket = connect(boundaries.get_index(metadata.hash_key));
+  send_message(socket.get(), &io_ops);
+
+  auto future = async(launch::async, [](unique_ptr<tcp::socket> socket) -> bool {
+      auto reply = read_reply<Reply> (socket.get());
+      socket->close();
+
+      if (reply->message != "TRUE") {
+      cerr << "[ERR] Failed to upload block . Details: " << reply->details << endl;
+      return false;
+      }
+
+      return true;
+      }, move(socket));
+
+  future.get();
+
   file_info.num_block = 1;
+  file_info.blocks_metadata.push_back(metadata);
+  file_info.uploading = 0;
 
+  socket = connect(file_hash_key);
   send_message(socket.get(), &file_info);
-  reply = read_reply<Reply> (socket.get());
+  auto reply = read_reply<Reply> (socket.get());
+
+  if (reply->message != "TRUE") {
+    cerr << "[ERR] Failed to upload file. Details: " << reply->details << endl;
+    return EXIT_FAILURE;
+  }
+
+  socket->close();
+
+  return EXIT_SUCCESS;
+}
+// }}}
+// write {{{
+uint64_t DFS::write(std::string& file_name, const char* buf, uint64_t off, uint64_t len) {
+  Histogram boundaries(NUM_NODES, 0);
+  boundaries.initialize();
+
+  //auto fd = get_file_description(
+    //std::bind(&connect, *this, std::placeholders::_1), file_name
+  //);
+
+  auto socket = connect(h(file_name));
+
+  FileRequest fr;
+  fr.name = file_name;
+
+  send_message(socket.get(), &fr);
+  auto fd = (read_reply<FileDescription> (socket.get()));
   socket->close();
+  if(fd == nullptr) return 0;
+
+  off = std::max(0ul, std::min(off, std::max(fd->size, BLOCK_SIZE - 1)));
+
+  //! Insert the blocks
+  vector<BlockMetadata> blocks_metadata;
+  vector<future<bool>> slave_sockets;
+
+  uint64_t to_write_bytes = len;
+  uint64_t written_bytes = 0;
+
+  int block_beg_seq = (int) off / BLOCK_SIZE;
+  int block_end_seq = (int) (len + off - 1) / BLOCK_SIZE;
+
+  for(int i=block_beg_seq; i<=block_end_seq; i++) {
```

Comment:
We need to cut the blocks in a new line for the mapreduce framework to successfully parse every record.